core/server/data/schema/schema.js

```diff
@@ -459,5 +459,31 @@ module.exports = {
         created_by: {type: 'string', maxlength: 24, nullable: false},
         updated_at: {type: 'dateTime', nullable: true},
         updated_by: {type: 'string', maxlength: 24, nullable: true}
+    },
+    email_batches: {
+        id: {type: 'string', maxlength: 24, nullable: false, primary: true},
+        email_id: {type: 'string', maxlength: 24, nullable: false, references: 'emails.id'},
+        provider_id: {type: 'string', maxlength: 255, nullable: true},
+        status: {
+            type: 'string',
+            maxlength: 50,
+            nullable: false,
+            defaultTo: 'pending',
+            validations: {isIn: [['pending', 'submitting', 'submitted', 'failed']]}
+        },
+        error_data: {type: 'text', maxlength: 1000000000, fieldtype: 'long', nullable: true}
+    },
+    email_recipients: {
+        id: {type: 'string', maxlength: 24, nullable: false, primary: true},
+        email_id: {type: 'string', maxlength: 24, nullable: false, references: 'emails.id'},
+        member_id: {type: 'string', maxlength: 24, nullable: false, index: true},
+        batch_id: {type: 'string', maxlength: 24, nullable: false, references: 'email_batches.id'},
+        processed_at: {type: 'dateTime', nullable: true},
+        delivered_at: {type: 'dateTime', nullable: true},
+        failed_at: {type: 'dateTime', nullable: true},
```

Comment:
Thinking this through, failed is going to come with a reason, and I'm not sure that storing error_data on a batch is right either if it's still a big JSON blob? 

I think we have 2 options here, either spend a bit more time thinking about how this will actually get used in order to get the structure right first time, or pull these columns until we're sure how we would use them? 

Having to add columns later is easy. Having to modify columns later is a PITA.

Reply:
> failed is going to come with a reason
Yeah, I thought that might be in whatever "activities" table we had later so that we have a history of attempts and associated errors, unless we plan to update this record based on events coming from Mailgun and only keep the latest error state until it's a success?

> I'm not sure that storing error_data on a batch is right either if it's still a big JSON blob
If it's just for auditing a JSON blob could be useful when there's no access to the logs. Seeing as this is an API error response that we want to store perhaps we should follow our own error pattern and have a message/details/context that the bulk email provider can populate by transforming the provider error?

Reply:
> I thought that might be in whatever "activities" table we had later so that we have a history of attempts and associated errors, unless we plan to update this record based on events coming from Mailgun and only keep the latest error state until it's a success?

This is IMO confusing permanent and temporary failures which are different. AFAIK there can be only one permanent fail per email.

> If it's just for auditing a JSON blob could be useful when there's no access to the logs. Otherwise, seeing as this is an API error response that we want to store perhaps we should follow our own error pattern and have a message/details/context that the bulk email provider can populate by transforming the provider error?

I think what's missing here is clarity on what the Mailgun API _can_ return?

Reply:
> This is IMO confusing permanent and temporary failures which are different. AFAIK there can be only one permanent fail per email.

I think there can only be one permanent failure unless the email is manually retried from the Mailgun UI? I was thinking more about exposing that delivery attempts have been made+failed/are continuing so that member-site support staff know what's happening. Was the intention for the email_recipient record to show failed only for a permanent failure?

> I think what's missing here is clarity on what the Mailgun API can return?

Agreed. A quick scan of their docs doesn't provide anything more helpful than a description of HTTP status codes 😞 I guess we'd also need to consider errors occurring before the Mailgun API endpoint is reached such as network issues, firewalls, gateway timeouts etc?

Reply:
All of this is a much much bigger discussion - so I suggest going with option 1 and pulling the fields we're not sure about yet.